src/utils/sources-tree/addToTree.js

```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
 import type { Node } from "./types";
 import type { SourceRecord } from "../../reducers/types";
 
-function createNodeInTree(part: string, path: string, tree: Node) {
+function createNodeInTree(
+  part: string,
+  path: string,
+  tree: Node,
+  index: number
+) {
   const node = createNode(part, path, []);
+
   // we are modifying the tree
-  tree.contents = [...tree.contents, node];
+  const contents = tree.contents.slice(0);
+  contents.splice(index, 0, node);
+  tree.contents = contents;
+
   return node;
 }
 
 /*
+ * Gets domain from url (without www prefix)
+ */
+function getDomain(url: string): ?string {
+  const { host } = parse(url);
+  if (!host) {
+    return null;
+  }
+  return host.indexOf("www.") === 0 ? host.substr(4) : host;
+}
+
+/*
+ * Checks if node name matches debugger host/domain.
+ */
+function isExactDomainMatch(part: string, debuggeeHost: string): boolean {
+  return part.indexOf("www.") === 0
+    ? part.substr(4) === debuggeeHost
+    : part === debuggeeHost;
+}
+
+type FindNodeInContentsMatcher = any => number;
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
+ * number, index of the found child, or negative number, which can be used
+ * to calculate a position where a new node can be inserted (`-index - 1`).
+ * The matcher is a function that returns result of comparision of a node with
+ * lookup value.
+ */
+function findNodeInContents(tree: Node, matcher: FindNodeInContentsMatcher) {
```

Comment:
describing this function in plain english, and using the ternary values I defined elsewhere (`T_TRUE = 1, T_UNKNOWN = 0, T_FALSE = -1`):

* we take the tree contents which is an array. if that array is empty, we return -1.

* if it is not empty, we set a start point ('l' for left) and an end point ('r' for right)

* so long as the starting point is less than the end point, we also select a center point ('c')

* if the center point is `T_FALSE` ( -1 ) -- then it should be at the bottom of the list, so we set the start point (l) to c + 1

* otherwise, the center point is `T_UNKNOWN`, or `T_TRUE`, so we set the ending point to C

example:

```
0|1|2|3|4
----------
l| |c| |r

if (matcher(contents[c]) === `T_FALSE`) 
0|1|2|3|4
----------
 | | |l, c|r // ignore the first half of the array

else
0|1|2|3|4
----------
l|c|r| |  // ignore the second half of the array
```

* Once the loop completes, (and our starting point "l" is greater than our endpoint "r") 

```
0|1|2|3|4
----------
 |r|l| |
0|1|2|3|4
----------
 | | | |r|l // outside of the original array if all matcher checks return `T_FALSE`
```

we check if the result is `T_UNKNOWN`. If it is unknown, we return the starting value ("l").

If it is either `T_FALSE` (-1) or `T_TRUE` (1), then 

- check if it is `T_FALSE`, if it is, return the starting point ('l'). 
- if it is `T_TRUE`, return the starting point ('l') plus 1. 

whatever the result is, make it negative and -1 from it.

```
0|1|2|3|4
----------
 |r|l|x|   // if `T_FALSE` is true, the returned position will be -(l + 1) ( also see above for the off by one case, where  | | | |r|l|x -- is outside of the existing array by two spots)
0|1|2|3|4
----------
 |r|l| |x| // if `T_TRUE` is true, the returned position will be -(l + 2)
```

this code is essentially taking an array, seeing if the matching position would be above or below the center, and halving it to the side where the element might be found. That isnt immediately clear from the code. there are two things contributing to this: 

1) bad naming. Single character variable names shouldn't be used except in special situations, and this isn't one of them.
2) return value is doing too much. it is unclear to me why we need it to be negative, or why we need to add +1 to it. can you clarify why this is happening? also, there are a lot of magic things happening, for example +1 is being added in a number of places without any reason explained behind it. return value is a good example of "why are we adding 1 and then making the whole thing negative and then adding 1 to it again". at the very least the return value needs to be split up


it also seems like we are avoiding complex data structures, but in this case i think we need them. we shouldnt be checking for negative indexs to tell if a node is new or not.

Reply:
Sorry, I don't follow your analysis. The algorithm above is basic binary search algorithm. I renamed 'l' and 'r' to 'left' and 'right' for clarity.

Reply:
> describing this function in plain english, and using the ternary values I defined elsewhere

The algorithm is using return values similar to comparefn defined in http://www.ecma-international.org/ecma-262/5.1/#sec-15.4.4.11 , and these are not compatible with  ternary values.

Reply:
:+1: i mean ternary as in: https://en.wikipedia.org/wiki/Three-valued_logic -- which i see as being similar to what is going on in the sort / comparefn that you linked (the difference being that 0 is used for "equal" instead of "undefined"). But i could be wrong... anyway, thank you for the changes it reads much better!

Reply:
there is no guarantee that localeCompare() will return only -1, 0, or 1 (based on spec). Writing `const result = name1.localeCompare(name2); return result === 0 ? T_UNKNOWN : result < 0 ? T_FALSE : T_TRUE;` to accommodate ternary values does not make any sense (and unreadable) IMHO.